### PR TITLE
Add `isSoftDeleted` helper method to model instances

### PR DESCRIPTION
**Pull request by jsatk**
### Pull Request check-list

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)? **I need to fix failing tests**
- [x] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Have you added an entry under `Future` in the changelog?

### Description of change

Since users can rename the `deletedAt` field when creating a model it is helpful to have a `isSoftDeleted` so other users of the DB don't have to hard code the custom `deletedAt` field name throughout their codebase.

Addresses: https://github.com/sequelize/sequelize/issues/7408

### CODE REVIEW

**Comment by jsatk** [comment]
The test suite seems non-deterministic. It was passing. Adore and I pushed copy-only changes and now it's failing. I'd love to help fix if possible but a bit stumped st the moment.

File: test/integration/instance.test.js [hunk 1]
```diff
@@ -2076,6 +2076,69 @@ describe(Support.getTestDialectTeaser('Instance'), () => {
     });
   });
 
+  describe('isSoftDeleted', () => {
+    beforeEach(function() {
+      this.ParanoidUser = this.sequelize.define('ParanoidUser', {
+        username: { type: DataTypes.STRING }
+      }, { paranoid: true });
+
+      this.ParanoidUser.hasOne(this.ParanoidUser);
```
Comment (janmeier) [inline test/integration/instance.test.js]: Is this relation needed?

File: test/unit/instance/is-soft-deleted.test.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    it('should not throw if paranoid is set to true', function() {
+      expect(() => {
+        this.paranoidUser.isSoftDeleted();
+      }).to.not.throw();
+    });
+
+    it('should throw if paranoid is set to false', function() {
+      expect(() => {
+        this.user.isSoftDeleted();
+      }).to.throw();
```
Comment (janmeier) [inline test/unit/instance/is-soft-deleted.test.js]: `to.throw('Model is not paranoid')`

File: test/integration/instance.test.js (hunk 1 above)
Comment (jsatk) [inline test/integration/instance.test.js]: I'm not entirely sure but I believe so. https://github.com/sequelize/sequelize/pull/7531/files/354210b1992afccaae8fdb270fddcb8e4752923f#diff-2756a0ac12fc81f7843fd569393b5966R1655 has the same code and I found that I needed that logic here.

File: test/unit/instance/is-soft-deleted.test.js (hunk 2 above)
Comment (jsatk) [inline test/unit/instance/is-soft-deleted.test.js]: Done! Will amend fix to this commit.

**Comment by jsatk** [comment]
@janmeier thank you! I responded to one and fixed the other. Thanks for the review. Excited to contribute. 😄

File: test/integration/instance.test.js [hunk 3]
```diff
@@ -2076,6 +2076,69 @@ describe(Support.getTestDialectTeaser('Instance'), () => {
...
+      return this.ParanoidUser.sync({ force: true });
+    });
+
+    it('returns false if user is not soft deleted', function() {
+      const self = this;
```
Comment (felixfbecker) [inline test/integration/instance.test.js]: Don't use `self`

File: lib/model.js [hunk 4]
```diff
@@ -3747,6 +3747,24 @@ class Model {
   }
 
   /**
+   * Helper method to determine if a instance is "soft deleted".  This is
+   * particularly useful if the implementer renamed the `deletedAt` attribute
+   * to something different.  This method requires `paranoid` to be enabled.
+   *
+   * @returns {Boolean}
+   */
+  isSoftDeleted() {
```
Comment (felixfbecker) [inline lib/model.js]: Yeah, shouldn't the function just check if `deletedAt` is set and less than current datetime? The default value logic looks weird to me

File: test/integration/instance.test.js [hunk 5]
```diff
@@ -2076,6 +2076,69 @@ describe(Support.getTestDialectTeaser('Instance'), () => {
     });
   });
 
+  describe('isSoftDeleted', () => {
```
Comment (jsatk) [inline test/integration/instance.test.js]: According to the linter in sequelize this is incorrect. It asks that you use arrow functions when possible over `function` and only use `function` when scope of `this` is a concern. 

`Unexpected function expression. (prefer-arrow-callback)`.

See https://github.com/sequelize/sequelize/blob/master/.eslintrc.json#L15

File: test/integration/instance.test.js (hunk 3 above)
Comment (jsatk) [inline test/integration/instance.test.js]: Good call. It's not needed due to arrow function.

File: lib/model.js (hunk 4 above)
Comment (jsatk) [inline lib/model.js]: @felixfbecker The logic is meant to get the value of the property being used for soft delete (which by default is `deletedAt`. 

@sushantdhiman I didn't consider future deletes. I will update the logic to return true if the value of the soft delete property is same as or before now.

File: lib/model.js [hunk 6]
```diff
@@ -3747,6 +3747,24 @@ class Model {
...
+    if (!this.constructor._timestampAttributes.deletedAt) {
+      throw new Error('Model is not paranoid');
+    }
+
+    const deletedAtAttribute = this.constructor.rawAttributes[this.constructor._timestampAttributes.deletedAt];
+    const notSoftDeleted = deletedAtAttribute.hasOwnProperty('defaultValue') ? deletedAtAttribute.defaultValue : null;
+
+    return this.getDataValue(this.constructor._timestampAttributes.deletedAt) !== notSoftDeleted;
```
Comment (jsatk) [inline lib/model.js]: @sushantdhiman Will do! What is the difference between `this.get` and `this.getDataValue`? I see both being used in the codebase.

File: test/integration/instance.test.js (hunk 1 above)
Comment (jsatk) [inline test/integration/instance.test.js]: I've removed it and all tests continue to pass. It seems it is not needed.

**Comment by jsatk** [comment]
@felixfbecker @sushantdhiman I've addressed all comments and added in a few more unit tests. Please let me know if there's anything else. Thanks for the review. :smile:

**Comment by janmeier** [comment]
@jsatk Fails for sqlite and mssql

``` 
  1) [SQLITE] Instance isSoftDeleted returns true if user is soft deleted:
     AssertionError: expected false to be true
      at null.<anonymous> (node_modules/chai/lib/chai/core/assertions.js:278:10)
      at addProperty (node_modules/chai/lib/chai/utils/addProperty.js:43:29)
      at doAsserterAsyncAndAddThen (node_modules/chai-as-promised/lib/chai-as-promised.js:296:33)
      at null.<anonymous> (node_modules/chai-as-promised/lib/chai-as-promised.js:286:25)
      at Object.defineProperty.get (node_modules/chai/lib/chai/utils/overwriteProperty.js:50:37)
      at test/integration/instance.test.js:2100:51
  From previous event:
      at Promise.then (node_modules/cls-bluebird/lib/shimMethod.js:38:20)
  2) [SQLITE] Instance isSoftDeleted works with custom `deletedAt` field name:
     AssertionError: expected false to be true
      at null.<anonymous> (node_modules/chai/lib/chai/core/assertions.js:278:10)
      at addProperty (node_modules/chai/lib/chai/utils/addProperty.js:43:29)
      at doAsserterAsyncAndAddThen (node_modules/chai-as-promised/lib/chai-as-promised.js:296:33)
      at null.<anonymous> (node_modules/chai-as-promised/lib/chai-as-promised.js:286:25)
      at Object.defineProperty.get (node_modules/chai/lib/chai/utils/overwriteProperty.js:50:37)
      at test/integration/instance.test.js:2127:53
  From previous event:
      at Promise.then (node_modules/cls-bluebird/lib/shimMethod.js:38:20)
```

**Comment by jsatk** [comment]
Apologies.

I saw that this failed the other day and haven't had time to fix.  Will hopefully fix sometime this week.

**Comment by jsatk** [comment]
@janmeier Any idea why the test suite date time is set to `1970-01-01T00:00:00.000Z`? Both my `deleted_at` and `now` values have a date time of `1970-01-01T00:00:00.000Z`. Any help? :)

File: lib/model.js [hunk 7]
```diff
@@ -3747,6 +3747,38 @@ class Model {
...
+   */
+  isSoftDeleted() {
+    if (!this.constructor._timestampAttributes.deletedAt) {
+      throw new Error('Model is not paranoid');
+    }
+
+    const deletedAtAttribute = this.constructor.rawAttributes[this.constructor._timestampAttributes.deletedAt];
+    const defaultValue = deletedAtAttribute.hasOwnProperty('defaultValue') ? deletedAtAttribute.defaultValue : null;
+    const deletedAt = this.get(this.constructor._timestampAttributes.deletedAt);
+    const isNotSet = deletedAt === defaultValue;
```
Comment (felixfbecker) [inline lib/model.js]: Avoid booleans with negatives, it's hard to read something like `!isNotSet`. Prefer `isSet`

Comment (jsatk) [inline lib/model.js]: Done!

**Comment by jsatk** [comment]
@felixfbecker @janmeier @sushantdhiman I believe this PR is good to go. Any other concerns?

**Pull request merged**